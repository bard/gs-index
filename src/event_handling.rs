use sea_query::{Expr, Iden, PostgresQueryBuilder, Query};
use tokio_postgres::{Client, Error, NoTls, Transaction};

pub const DB_SCHEMA: &str = r#"
   CREATE TABLE project (
     chain_id INTEGER NOT NULL,
     project_id VARCHAR NOT NULL,
     created_at_block BIGINT NOT NULL,
     metadata JSONB,
     PRIMARY KEY(chain_id, project_id)
   );
"#;

#[derive(Iden)]
enum Project {
    Table,
    ChainId,
    ProjectId,
    CreatedAtBlock,
    Metadata,
    Owners,
}

#[derive(Iden)]
enum Round {
    Table,
    ChainId,
    RoundAddress,
    CreatedAtBlock,
}

// TODO use stricter types
pub struct Event {
    pub chain_id: i32,
    pub address: String,
    pub block_number: i32,
    pub payload: EventPayload,
}

pub enum EventPayload {
    ProjectCreated {
        project_id: String,
    },
    MetadataUpdated {
        project_id: String,
        meta_ptr: MetaPtr,
    },
    OwnerAdded {
        project_id: String,
        owner: String,
    },
    OwnerRemoved {
        project_id: String,
        owner: String,
    },
    RoundCreated {
        round_address: String,
    },
}

pub struct MetaPtr {
    pub pointer: String,
}

pub type IpfsGetter = fn(&String) -> String;

pub struct ChangeSet {
    pub sql: String,
}

pub fn event_to_changeset(event: &Event, ipfs: IpfsGetter) -> ChangeSet {
    match &event.payload {
        EventPayload::ProjectCreated { project_id } => ChangeSet {
            sql: Query::insert()
                .into_table(Project::Table)
                .columns([
                    Project::ChainId,
                    Project::ProjectId,
                    Project::CreatedAtBlock,
                ])
                .values_panic([
                    event.chain_id.into(),
                    project_id.into(),
                    event.block_number.into(),
                ])
                .to_string(PostgresQueryBuilder),
        },

        EventPayload::MetadataUpdated {
            meta_ptr,
            project_id,
        } => {
            let metadata = ipfs(&meta_ptr.pointer);
            ChangeSet {
                sql: Query::update()
                    .table(Project::Table)
                    .values([(Project::Metadata, metadata.into())])
                    .and_where(Expr::col(Project::ChainId).eq(event.chain_id))
                    .and_where(Expr::col(Project::ProjectId).eq(project_id))
                    .to_string(PostgresQueryBuilder),
            }
        }

        EventPayload::OwnerAdded { project_id, owner } => {
            ChangeSet {
                // TODO build SQL safely
                sql: format!(
                    r#"UPDATE "project" SET "owners" = ("owners" || '["{}"]') WHERE "chain_id" = {} AND "project_id" = '{}'"#,
                    owner, event.chain_id, project_id
                ),
            }
        }

        EventPayload::OwnerRemoved { project_id, owner } => {
            ChangeSet {
                // TODO build SQL safely
                sql: format!(
                    r#"UPDATE "project" SET "owners" = ("owners" - '{}') WHERE "chain_id" = {} AND "project_id" = '{}'"#,
                    owner, event.chain_id, project_id
                ),
            }
        }

        EventPayload::RoundCreated { round_address } => ChangeSet {
            sql: Query::insert()
                .into_table(Round::Table)
                .columns([Round::ChainId, Round::RoundAddress, Round::CreatedAtBlock])
                .values_panic([
                    event.chain_id.into(),
                    round_address.into(),
                    event.block_number.into(),
                ])
                .to_string(PostgresQueryBuilder),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn dummy_ipfs_getter(_cid: &String) -> String {
        return "".to_string();
    }

    #[test]
    fn test_handle_project_created() {
        let event = Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            payload: EventPayload::ProjectCreated {
                project_id: "proj-123".to_string(),
            },
        };

        assert_eq!(
            event_to_changeset(&event, dummy_ipfs_getter).sql,
            r#"INSERT INTO "project" ("chain_id", "project_id", "created_at_block") VALUES (1, 'proj-123', 4242)"#
        );
    }

    #[test]
    fn test_handle_metadata_updated() {
        let event = Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            payload: EventPayload::MetadataUpdated {
                project_id: "proj-123".to_string(),
                meta_ptr: MetaPtr {
                    pointer: "123".to_string(),
                },
            },
        };
        let ipfs_getter = |_cid: &String| -> String { r#"{ "foo": "bar" }"#.to_string() };

        assert_eq!(
            event_to_changeset(&event, ipfs_getter).sql,
            r#"UPDATE "project" SET "metadata" = E'{ \"foo\": \"bar\" }' WHERE "chain_id" = 1 AND "project_id" = 'proj-123'"#
        );
    }

    #[test]
    fn test_handle_owner_added() {
        let event = Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            payload: EventPayload::OwnerAdded {
                project_id: "proj-123".to_string(),
                owner: "0x123".to_string(),
            },
        };

        assert_eq!(
            event_to_changeset(&event, dummy_ipfs_getter).sql,
            r#"UPDATE "project" SET "owners" = ("owners" || '["0x123"]') WHERE "chain_id" = 1 AND "project_id" = 'proj-123'"#
        );
    }

    #[test]
    fn test_handle_owner_removed() {
        let event = Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            payload: EventPayload::OwnerRemoved {
                project_id: "proj-123".to_string(),
                owner: "0x123".to_string(),
            },
        };

        assert_eq!(
            event_to_changeset(&event, dummy_ipfs_getter).sql,
            r#"UPDATE "project" SET "owners" = ("owners" - '0x123') WHERE "chain_id" = 1 AND "project_id" = 'proj-123'"#
        );
    }

    #[test]
    fn test_handle_contract_round_created() {
        let event = Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            payload: EventPayload::RoundCreated {
                round_address: "0x123".to_string(),
            },
        };

        let ChangeSet { sql } = event_to_changeset(&event, dummy_ipfs_getter);
        assert_eq!(
            sql,
            r#"INSERT INTO "round" ("chain_id", "round_address", "created_at_block") VALUES (1, '0x123', 4242)"#
        );
    }
}
