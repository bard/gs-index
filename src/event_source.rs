use async_stream::stream;
use futures::pin_mut;
use futures::stream::Stream;
use futures::stream::StreamExt;
use serde_json::{from_str, to_string};
use std::fs::File;
use std::io::Write;
use std::io::{self, BufRead, BufReader};
use std::path::Path;

use crate::event_handling::{Event, EventPayload};

pub fn event_stream_from_vector(
    events: Vec<Event>,
    start: usize,
) -> impl Stream<Item = (Event, usize)> {
    let mut index = 0;
    stream! {
        for event in events {
            if index >= start {
                yield (event, index)
            }
            index += 1
        }
    }
}

pub fn event_stream_from_ndjson_file(
    file: File,
    start: usize,
    warn_on_unparseable_items: bool,
) -> impl Stream<Item = (Event, usize)> {
    event_stream_from_buf_reader(io::BufReader::new(file), start, warn_on_unparseable_items)
}

pub fn event_stream_from_ndjson_stdin(
    start: usize,
    warn_on_unparseable_items: bool,
) -> impl Stream<Item = (Event, usize)> {
    event_stream_from_buf_reader(
        io::BufReader::new(io::stdin()),
        start,
        warn_on_unparseable_items,
    )
}

pub fn event_stream_from_buf_reader<R: io::BufRead>(
    reader: R,
    start: usize,
    warn_on_unparseable_items: bool,
) -> impl Stream<Item = (Event, usize)> {
    let mut index = 0;
    stream! {
        for line in reader.lines() {
            if index >= start {
                let line_content = line.unwrap();
                let parse_result: Result<Event, _>  = serde_json::from_str(&line_content);
                match parse_result {
                    Ok(event) => {
                        yield (event, index);
                    }
                    Err(err) => {
                        // TODO propagate error instead of panicking. Might need to
                        // use try_stream!
                        if warn_on_unparseable_items {
                            eprintln!("Warning: skipping event due to parse error: {}. Data: {}", err, line_content);
                        }
                    }
                }
            }
            index += 1;
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs::OpenOptions;

    use tempfile::NamedTempFile;

    use super::*;
    use crate::event_handling::{EventPayload, MetaPtr};

    #[tokio::test]
    async fn test_event_stream_from_ndjson_file() {
        let mut tmp_file = NamedTempFile::new().unwrap();
        for event in get_sample_events() {
            writeln!(tmp_file, "{}", to_string(&event).unwrap()).unwrap();
        }
        let file_name = tmp_file.path();

        let file_event_source =
            event_stream_from_ndjson_file(File::open(file_name).unwrap(), 0, false);
        pin_mut!(file_event_source);

        let (event, index) = file_event_source.next().await.unwrap();
        assert_eq!(index, 0);
        assert_eq!(event.block_number, 10);
    }

    #[tokio::test]
    async fn test_event_stream_from_vector() {
        let event_data = get_sample_events();
        let in_memory_event_source = event_stream_from_vector(event_data, 0);
        pin_mut!(in_memory_event_source);

        let (event, index) = in_memory_event_source.next().await.unwrap();
        assert_eq!(index, 0);
        assert_eq!(event.block_number, 10);

        let (event, index) = in_memory_event_source.next().await.unwrap();
        assert_eq!(index, 1);
        assert_eq!(event.block_number, 20);

        let (event, index) = in_memory_event_source.next().await.unwrap();
        assert_eq!(index, 2);
        assert_eq!(event.block_number, 30);
    }

    #[tokio::test]
    async fn test_events_stream_from_vector_starting_at_index() {
        let event_data = get_sample_events();
        let in_memory_event_source = event_stream_from_vector(event_data, 2);
        pin_mut!(in_memory_event_source);

        let (event, index) = in_memory_event_source.next().await.unwrap();
        assert_eq!(index, 2);
        assert_eq!(event.block_number, 30);
    }

    fn get_sample_events() -> Vec<Event> {
        vec![
            Event {
                chain_id: 1,
                address: "0x123".to_string(),
                block_number: 10,
                log_index: 0,
                data: EventPayload::ProjectCreated {
                    project_id: "proj-123".to_string(),
                },
            },
            Event {
                chain_id: 1,
                address: "0x123".to_string(),
                block_number: 20,
                log_index: 1,
                data: EventPayload::MetadataUpdated {
                    project_id: "proj-123".to_string(),
                    meta_ptr: MetaPtr {
                        pointer: "123".to_string(),
                    },
                },
            },
            Event {
                chain_id: 1,
                address: "0x123".to_string(),
                block_number: 30,
                log_index: 1,
                data: EventPayload::OwnerAdded {
                    project_id: "proj-123".to_string(),
                    owner: "0x123".to_string(),
                },
            },
        ]
    }
}
