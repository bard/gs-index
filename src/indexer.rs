use std::pin::Pin;

use crate::event_handling::{
    event_to_changeset, events_to_change_sets_sequential, ChangeSet, Event, EventPayload,
    IpfsGetter, MetaPtr, DB_SCHEMA,
};
use async_stream::stream;
use futures::pin_mut;
use futures::stream::Stream;
use futures::stream::StreamExt;
use sea_query::{Expr, Iden, PostgresQueryBuilder, Query};
use tokio_postgres::{Client, Connection, Error, NoTls, Transaction};

#[cfg(test)]
mod tests {
    use super::*;
    use crate::event_source::event_stream_from_vector;

    #[tokio::test]
    async fn test_project_created() {
        let ipfs_getter = |_cid: &str| -> String { "".to_string() };
        let events = vec![Event {
            chain_id: 1,
            address: "0x123".to_string(),
            block_number: 4242,
            log_index: 1,
            payload: EventPayload::ProjectCreated {
                project_id: "proj-123".to_string(),
            },
        }];

        let db_dump = event_stream_to_db_dump(events, ipfs_getter).await.unwrap();

        insta::assert_yaml_snapshot!(db_dump);
    }

    #[tokio::test]
    async fn test_project_created_and_metadata_updated() {
        let ipfs_getter = |_cid: &str| -> String { "{ \"foo\": \"bar\" }".to_string() };
        let events = vec![
            Event {
                chain_id: 1,
                address: "0x123".to_string(),
                block_number: 4242,
                log_index: 1,
                payload: EventPayload::ProjectCreated {
                    project_id: "proj-123".to_string(),
                },
            },
            Event {
                chain_id: 1,
                address: "0x123".to_string(),
                block_number: 4242,
                log_index: 2,
                payload: EventPayload::MetadataUpdated {
                    project_id: "proj-123".to_string(),
                    meta_ptr: MetaPtr {
                        pointer: "123".to_string(),
                    },
                },
            },
        ];

        let db_dump = event_stream_to_db_dump(events, ipfs_getter).await.unwrap();

        insta::assert_yaml_snapshot!(db_dump);
    }

    fn dummy_ipfs_getter(_cid: &str) -> String {
        "".into()
    }

    async fn event_stream_to_db_dump(
        events: Vec<Event>,
        ipfs_getter: IpfsGetter,
    ) -> Result<String, Error> {
        let connection_string = "host=localhost user=postgres password=postgres";
        let (mut client, connection) = tokio_postgres::connect(connection_string, NoTls).await?;
        tokio::spawn(connection);
        let transaction = client.transaction().await?;
        transaction.batch_execute(DB_SCHEMA).await?;

        let event_stream = event_stream_from_vector(events, 0);
        pin_mut!(event_stream);

        while let Some((event, _index)) = event_stream.next().await {
            let change_set = event_to_changeset(&event, ipfs_getter);
            transaction.simple_query(&change_set.sql).await?;
        }

        let rows = transaction
            .query("SELECT JSON_AGG(project) #>> '{}' FROM project;", &[])
            .await?;
        Ok(rows[0].get(0))
    }
}
