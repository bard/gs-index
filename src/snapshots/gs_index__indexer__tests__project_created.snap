---
source: src/indexer.rs
expression: db_dump
---
"[{\"chain_id\":1,\"project_id\":\"proj-123\",\"created_at_block\":4242,\"metadata\":null}]"

